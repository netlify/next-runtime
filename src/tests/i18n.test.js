// Test next-on-netlify when i18n is set in next.config.js (Next 10+)

const { parse, join } = require('path')
const { existsSync, readdirSync, readFileSync, readJsonSync } = require('fs-extra')
const buildNextApp = require('./helpers/buildNextApp')

// The name of this test file (without extension)
const FILENAME = parse(__filename).name

// The directory which will be used for testing.
// We simulate a NextJS app within that directory, with pages, and a
// package.json file.
const PROJECT_PATH = join(__dirname, 'builds', FILENAME)

const DEFAULT_LOCALE = 'en'

// Capture the output to verify successful build
let buildOutput

beforeAll(
  async () => {
    buildOutput = await buildNextApp()
      .forTest(__filename)
      .withPages('pages')
      .withNextConfig('next.config.js-with-i18n.js')
      .withPackageJson('package.json')
      .build()
  },
  // time out after 180 seconds
  180 * 1000,
)

describe('next-on-netlify', () => {
  test('builds successfully', () => {
    expect(buildOutput).toMatch('Next on Netlify')
    expect(buildOutput).toMatch('Success! All done!')
  })
})

describe('next-on-netlify', () => {
  test('builds successfully', () => {
    expect(buildOutput).toMatch('Next on Netlify')
    expect(buildOutput).toMatch(`Copying static NextJS assets to out_publish`)
    expect(buildOutput).toMatch(`Setting up API endpoints as Netlify Functions in out_functions`)
    expect(buildOutput).toMatch(`Setting up pages with getInitialProps as Netlify Functions in out_functions`)
    expect(buildOutput).toMatch(`Setting up pages with getServerSideProps as Netlify Functions in out_functions`)
    expect(buildOutput).toMatch(`Copying pre-rendered pages with getStaticProps and JSON data to out_publish`)
    expect(buildOutput).toMatch(
      `Setting up pages with getStaticProps and fallback: true as Netlify Functions in out_functions`,
    )
    expect(buildOutput).toMatch(
      `Setting up pages with getStaticProps and revalidation interval as Netlify Functions in out_functions`,
    )
    expect(buildOutput).toMatch(`Copying pre-rendered pages without props to out_publish`)
    expect(buildOutput).toMatch('Setting up redirects')
    expect(buildOutput).toMatch('Success! All done!')
  })
})

describe('SSR Pages', () => {
  const functionsDir = join(PROJECT_PATH, 'out_functions')

  test('creates a Netlify Function for each SSR page', () => {
    expect(existsSync(join(functionsDir, 'next_index', 'next_index.ts'))).toBe(true)
    expect(existsSync(join(functionsDir, 'next_shows_id', 'next_shows_id.ts'))).toBe(true)
    expect(existsSync(join(functionsDir, 'next_shows_params', 'next_shows_params.ts'))).toBe(true)
    expect(existsSync(join(functionsDir, 'next_getServerSideProps_static', 'next_getServerSideProps_static.ts'))).toBe(
      true,
    )
    expect(existsSync(join(functionsDir, 'next_getServerSideProps_id', 'next_getServerSideProps_id.ts'))).toBe(true)
  })
})

describe('API Pages', () => {
  const functionsDir = join(PROJECT_PATH, 'out_functions')

  test('creates a Netlify Function for each API endpoint', () => {
    expect(existsSync(join(functionsDir, 'next_api_static', 'next_api_static.ts'))).toBe(true)
    expect(existsSync(join(functionsDir, 'next_api_shows_id', 'next_api_shows_id.ts'))).toBe(true)
    expect(existsSync(join(functionsDir, 'next_api_shows_params', 'next_api_shows_params.ts'))).toBe(true)
  })
})

describe('SSG Pages with getStaticProps', () => {
  test('creates pre-rendered HTML file in output directory', () => {
    const OUTPUT_PATH = join(PROJECT_PATH, 'out_publish', DEFAULT_LOCALE)

    expect(existsSync(join(OUTPUT_PATH, 'getStaticProps', 'static.html'))).toBe(true)
    expect(existsSync(join(OUTPUT_PATH, 'getStaticProps', '1.html'))).toBe(true)
    expect(existsSync(join(OUTPUT_PATH, 'getStaticProps', '2.html'))).toBe(true)
    expect(existsSync(join(OUTPUT_PATH, 'getStaticProps', 'withFallback', '3.html'))).toBe(true)
    expect(existsSync(join(OUTPUT_PATH, 'getStaticProps', 'withFallback', '4.html'))).toBe(true)
    expect(existsSync(join(OUTPUT_PATH, 'getStaticProps', 'withFallback', 'my', 'path', '1.html'))).toBe(true)
    expect(existsSync(join(OUTPUT_PATH, 'getStaticProps', 'withFallback', 'my', 'path', '2.html'))).toBe(true)
  })

  test('creates data .json file in /_next/data/ directory', () => {
    // Get path to data files
    const dirs = readdirSync(join(PROJECT_PATH, 'out_publish', '_next', 'data'))
    expect(dirs.length).toBe(1)
    const dataDir = join(PROJECT_PATH, 'out_publish', '_next', 'data', dirs[0], DEFAULT_LOCALE)

    expect(existsSync(join(dataDir, 'getStaticProps', 'static.json'))).toBe(true)
    expect(existsSync(join(dataDir, 'getStaticProps', '1.json'))).toBe(true)
    expect(existsSync(join(dataDir, 'getStaticProps', '2.json'))).toBe(true)
    expect(existsSync(join(dataDir, 'getStaticProps', 'withFallback', '3.json'))).toBe(true)
    expect(existsSync(join(dataDir, 'getStaticProps', 'withFallback', '4.json'))).toBe(true)
    expect(existsSync(join(dataDir, 'getStaticProps', 'withFallback', 'my', 'path', '1.json'))).toBe(true)
    expect(existsSync(join(dataDir, 'getStaticProps', 'withFallback', 'my', 'path', '2.json'))).toBe(true)
  })

  test('creates Netlify Functions for pages with fallback', () => {
    const functionPath1 = 'next_getStaticProps_withFallback_id/next_getStaticProps_withFallback_id.ts'
    expect(existsSync(join(PROJECT_PATH, 'out_functions', functionPath1))).toBe(true)

    const functionPath2 = 'next_getStaticProps_withFallback_slug/next_getStaticProps_withFallback_slug.ts'
    expect(existsSync(join(PROJECT_PATH, 'out_functions', functionPath2))).toBe(true)
  })
})

describe('SSG Pages with getStaticProps and revalidate', () => {
  const functionsDir = join(PROJECT_PATH, 'out_functions')

  test('creates a Netlify Function for each page', () => {
    expect(
      existsSync(join(functionsDir, 'next_getStaticProps_withrevalidate', 'next_getStaticProps_withrevalidate.ts')),
    ).toBe(true)
    expect(
      existsSync(
        join(functionsDir, 'next_getStaticProps_withRevalidate_id', 'next_getStaticProps_withRevalidate_id.ts'),
      ),
    ).toBe(true)
    expect(
      existsSync(
        join(
          functionsDir,
          'next_getStaticProps_withRevalidate_withFallback_id',
          'next_getStaticProps_withRevalidate_withFallback_id.ts',
        ),
      ),
    ).toBe(true)
  })
})

describe('Static Pages', () => {
  test('copies static pages to output directory', () => {
    const OUTPUT_PATH = join(PROJECT_PATH, 'out_publish', DEFAULT_LOCALE)

    expect(existsSync(join(OUTPUT_PATH, 'static.html'))).toBe(true)
    expect(existsSync(join(OUTPUT_PATH, 'static/[id].html'))).toBe(true)
  })

  test('copies static assets to out_publish/_next/ directory', () => {
    const dirs = readdirSync(join(PROJECT_PATH, 'out_publish', '_next', 'static'))

    expect(dirs.length).toBe(2)
    expect(dirs).toContain('chunks')
  })

  test.each(['en', 'es'])('multiple locales', (locale) => {
    const OUTPUT_PATH = join(PROJECT_PATH, 'out_publish', locale)

    expect(existsSync(join(OUTPUT_PATH, 'static.html'))).toBe(true)
    expect(existsSync(join(OUTPUT_PATH, 'static/[id].html'))).toBe(true)
  })

  test('locale not included in config', () => {
    const notIncludedLocale = 'fr'
    const OUTPUT_PATH = join(PROJECT_PATH, 'out_publish', notIncludedLocale)

    expect(existsSync(join(OUTPUT_PATH, 'static.html'))).toBe(false)
    expect(existsSync(join(OUTPUT_PATH, 'static/[id].html'))).toBe(false)
  })
})

describe('404 Page', () => {
  test('copies 404.html to output directory', () => {
    const OUTPUT_PATH = join(PROJECT_PATH, 'out_publish', DEFAULT_LOCALE)

    expect(existsSync(join(OUTPUT_PATH, '404.html'))).toBe(true)
  })
})

describe('Routing', () => {
  test('creates Netlify redirects', async () => {
    // Read _redirects file
    const contents = readFileSync(join(PROJECT_PATH, 'out_publish', '_redirects'))
    let redirects = contents.toString()

    // Replace non-persistent build ID with placeholder

    redirects = redirects.replace(/\/_next\/data\/[^\/]+\//g, '/_next/data/%BUILD_ID%/')

    // Check that redirects match
    expect(redirects).toMatchSnapshot()
  })
})
