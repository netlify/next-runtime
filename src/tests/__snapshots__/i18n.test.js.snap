// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routing creates Netlify redirects 1`] = `
"# Next-on-Netlify Redirects
/  /.netlify/functions/next_index  200
/404  /en/404.html  200
/_next/data/%BUILD_ID%/en.json  /.netlify/functions/next_index  200
/_next/data/%BUILD_ID%/en/getServerSideProps/static.json  /.netlify/functions/next_getServerSideProps_static  200
/_next/data/%BUILD_ID%/en/getStaticProps/1.json  /.netlify/functions/next_getStaticProps_id  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/en/getStaticProps/2.json  /.netlify/functions/next_getStaticProps_id  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/en/getStaticProps/static.json  /.netlify/functions/next_getStaticProps_static  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/en/getStaticProps/with-revalidate.json  /.netlify/functions/next_getStaticProps_withrevalidate  200
/_next/data/%BUILD_ID%/en/getStaticProps/withFallback/3.json  /.netlify/functions/next_getStaticProps_withFallback_id  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/en/getStaticProps/withFallback/4.json  /.netlify/functions/next_getStaticProps_withFallback_id  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/en/getStaticProps/withFallback/my/path/1.json  /.netlify/functions/next_getStaticProps_withFallback_slug  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/en/getStaticProps/withFallback/my/path/2.json  /.netlify/functions/next_getStaticProps_withFallback_slug  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/en/getStaticProps/withFallbackBlocking/3.json  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/en/getStaticProps/withFallbackBlocking/4.json  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/en/getStaticProps/withRevalidate/1.json  /.netlify/functions/next_getStaticProps_withRevalidate_id  200
/_next/data/%BUILD_ID%/en/getStaticProps/withRevalidate/2.json  /.netlify/functions/next_getStaticProps_withRevalidate_id  200
/_next/data/%BUILD_ID%/es.json  /.netlify/functions/next_index  200
/_next/data/%BUILD_ID%/es/getServerSideProps/static.json  /.netlify/functions/next_getServerSideProps_static  200
/_next/data/%BUILD_ID%/es/getStaticProps/static.json  /.netlify/functions/next_getStaticProps_static  200!  Cookie=__prerender_bypass,__next_preview_data
/_next/data/%BUILD_ID%/es/getStaticProps/with-revalidate.json  /.netlify/functions/next_getStaticProps_withrevalidate  200
/_next/data/%BUILD_ID%/getServerSideProps/static.json  /.netlify/functions/next_getServerSideProps_static  200
/api/hello-background  /.netlify/functions/next_api_hello-background  200
/api/static  /.netlify/functions/next_api_static  200
/en  /.netlify/functions/next_index  200
/en/getServerSideProps/static  /.netlify/functions/next_getServerSideProps_static  200
/en/getStaticProps/1  /.netlify/functions/next_getStaticProps_id  200!  Cookie=__prerender_bypass,__next_preview_data
/en/getStaticProps/2  /.netlify/functions/next_getStaticProps_id  200!  Cookie=__prerender_bypass,__next_preview_data
/en/getStaticProps/static  /.netlify/functions/next_getStaticProps_static  200!  Cookie=__prerender_bypass,__next_preview_data
/en/getStaticProps/with-revalidate  /.netlify/functions/next_getStaticProps_withrevalidate  200
/en/getStaticProps/withFallback/3  /.netlify/functions/next_getStaticProps_withFallback_id  200!  Cookie=__prerender_bypass,__next_preview_data
/en/getStaticProps/withFallback/4  /.netlify/functions/next_getStaticProps_withFallback_id  200!  Cookie=__prerender_bypass,__next_preview_data
/en/getStaticProps/withFallback/my/path/1  /.netlify/functions/next_getStaticProps_withFallback_slug  200!  Cookie=__prerender_bypass,__next_preview_data
/en/getStaticProps/withFallback/my/path/2  /.netlify/functions/next_getStaticProps_withFallback_slug  200!  Cookie=__prerender_bypass,__next_preview_data
/en/getStaticProps/withFallbackBlocking/3  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200!  Cookie=__prerender_bypass,__next_preview_data
/en/getStaticProps/withFallbackBlocking/4  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200!  Cookie=__prerender_bypass,__next_preview_data
/en/getStaticProps/withRevalidate/1  /.netlify/functions/next_getStaticProps_withRevalidate_id  200
/en/getStaticProps/withRevalidate/2  /.netlify/functions/next_getStaticProps_withRevalidate_id  200
/es  /.netlify/functions/next_index  200
/es/getServerSideProps/static  /.netlify/functions/next_getServerSideProps_static  200
/es/getStaticProps/static  /.netlify/functions/next_getStaticProps_static  200!  Cookie=__prerender_bypass,__next_preview_data
/es/getStaticProps/with-revalidate  /.netlify/functions/next_getStaticProps_withrevalidate  200
/getServerSideProps/static  /.netlify/functions/next_getServerSideProps_static  200
/getStaticProps/1  /.netlify/functions/next_getStaticProps_id  200!  Cookie=__prerender_bypass,__next_preview_data
/getStaticProps/1  /en/getStaticProps/1  200
/getStaticProps/2  /.netlify/functions/next_getStaticProps_id  200!  Cookie=__prerender_bypass,__next_preview_data
/getStaticProps/2  /en/getStaticProps/2  200
/getStaticProps/static  /.netlify/functions/next_getStaticProps_static  200!  Cookie=__prerender_bypass,__next_preview_data
/getStaticProps/static  /en/getStaticProps/static  200
/getStaticProps/with-revalidate  /.netlify/functions/next_getStaticProps_withrevalidate  200
/getStaticProps/withFallback/3  /.netlify/functions/next_getStaticProps_withFallback_id  200!  Cookie=__prerender_bypass,__next_preview_data
/getStaticProps/withFallback/3  /en/getStaticProps/withFallback/3  200
/getStaticProps/withFallback/4  /.netlify/functions/next_getStaticProps_withFallback_id  200!  Cookie=__prerender_bypass,__next_preview_data
/getStaticProps/withFallback/4  /en/getStaticProps/withFallback/4  200
/getStaticProps/withFallback/my/path/1  /.netlify/functions/next_getStaticProps_withFallback_slug  200!  Cookie=__prerender_bypass,__next_preview_data
/getStaticProps/withFallback/my/path/1  /en/getStaticProps/withFallback/my/path/1  200
/getStaticProps/withFallback/my/path/2  /.netlify/functions/next_getStaticProps_withFallback_slug  200!  Cookie=__prerender_bypass,__next_preview_data
/getStaticProps/withFallback/my/path/2  /en/getStaticProps/withFallback/my/path/2  200
/getStaticProps/withFallbackBlocking/3  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200!  Cookie=__prerender_bypass,__next_preview_data
/getStaticProps/withFallbackBlocking/3  /en/getStaticProps/withFallbackBlocking/3  200
/getStaticProps/withFallbackBlocking/4  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200!  Cookie=__prerender_bypass,__next_preview_data
/getStaticProps/withFallbackBlocking/4  /en/getStaticProps/withFallbackBlocking/4  200
/getStaticProps/withRevalidate/1  /.netlify/functions/next_getStaticProps_withRevalidate_id  200
/getStaticProps/withRevalidate/2  /.netlify/functions/next_getStaticProps_withRevalidate_id  200
/static  /en/static.html  200
/_next/data/%BUILD_ID%/en/getServerSideProps/all.json  /.netlify/functions/next_getServerSideProps_all_slug  200
/_next/data/%BUILD_ID%/en/getServerSideProps/all/*  /.netlify/functions/next_getServerSideProps_all_slug  200
/_next/data/%BUILD_ID%/en/getServerSideProps/:id.json  /.netlify/functions/next_getServerSideProps_id  200
/_next/data/%BUILD_ID%/en/getStaticProps/withFallback/:id.json  /.netlify/functions/next_getStaticProps_withFallback_id  200
/_next/data/%BUILD_ID%/en/getStaticProps/withFallback/:slug/*  /.netlify/functions/next_getStaticProps_withFallback_slug  200
/_next/data/%BUILD_ID%/en/getStaticProps/withFallbackBlocking/:id.json  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200
/_next/data/%BUILD_ID%/en/getStaticProps/withRevalidate/withFallback/:id.json  /.netlify/functions/next_getStaticProps_withRevalidate_withFallback_id  200
/_next/data/%BUILD_ID%/en/shows/:id.json  /.netlify/functions/next_shows_id  200
/_next/data/%BUILD_ID%/en/shows/:params/*  /.netlify/functions/next_shows_params  200
/_next/data/%BUILD_ID%/es/getServerSideProps/all.json  /.netlify/functions/next_getServerSideProps_all_slug  200
/_next/data/%BUILD_ID%/es/getServerSideProps/all/*  /.netlify/functions/next_getServerSideProps_all_slug  200
/_next/data/%BUILD_ID%/es/getServerSideProps/:id.json  /.netlify/functions/next_getServerSideProps_id  200
/_next/data/%BUILD_ID%/es/getStaticProps/withFallback/:id.json  /.netlify/functions/next_getStaticProps_withFallback_id  200
/_next/data/%BUILD_ID%/es/getStaticProps/withFallback/:slug/*  /.netlify/functions/next_getStaticProps_withFallback_slug  200
/_next/data/%BUILD_ID%/es/getStaticProps/withFallbackBlocking/:id.json  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200
/_next/data/%BUILD_ID%/es/getStaticProps/withRevalidate/withFallback/:id.json  /.netlify/functions/next_getStaticProps_withRevalidate_withFallback_id  200
/_next/data/%BUILD_ID%/es/shows/:id.json  /.netlify/functions/next_shows_id  200
/_next/data/%BUILD_ID%/es/shows/:params/*  /.netlify/functions/next_shows_params  200
/_next/data/%BUILD_ID%/getServerSideProps/all.json  /.netlify/functions/next_getServerSideProps_all_slug  200
/_next/data/%BUILD_ID%/getServerSideProps/all/*  /.netlify/functions/next_getServerSideProps_all_slug  200
/_next/data/%BUILD_ID%/getServerSideProps/:id.json  /.netlify/functions/next_getServerSideProps_id  200
/_next/data/%BUILD_ID%/getStaticProps/withFallback/:id.json  /.netlify/functions/next_getStaticProps_withFallback_id  200
/_next/data/%BUILD_ID%/getStaticProps/withFallback/:slug/*  /.netlify/functions/next_getStaticProps_withFallback_slug  200
/_next/data/%BUILD_ID%/getStaticProps/withFallbackBlocking/:id.json  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200
/_next/data/%BUILD_ID%/getStaticProps/withRevalidate/withFallback/:id.json  /.netlify/functions/next_getStaticProps_withRevalidate_withFallback_id  200
/_next/image*  url=:url w=:width q=:quality  /.netlify/functions/next_image?url=:url&w=:width&q=:quality  200
/api/shows/:id  /.netlify/functions/next_api_shows_id  200
/api/shows/:params/*  /.netlify/functions/next_api_shows_params  200
/en/getServerSideProps/all  /.netlify/functions/next_getServerSideProps_all_slug  200
/en/getServerSideProps/all/*  /.netlify/functions/next_getServerSideProps_all_slug  200
/en/getServerSideProps/:id  /.netlify/functions/next_getServerSideProps_id  200
/en/getStaticProps/withFallback/:id  /.netlify/functions/next_getStaticProps_withFallback_id  200
/en/getStaticProps/withFallback/:slug/*  /.netlify/functions/next_getStaticProps_withFallback_slug  200
/en/getStaticProps/withFallbackBlocking/:id  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200
/en/getStaticProps/withRevalidate/withFallback/:id  /.netlify/functions/next_getStaticProps_withRevalidate_withFallback_id  200
/en/shows/:id  /.netlify/functions/next_shows_id  200
/en/shows/:params/*  /.netlify/functions/next_shows_params  200
/en/static/:id  /en/static/[id].html  200
/es/getServerSideProps/all  /.netlify/functions/next_getServerSideProps_all_slug  200
/es/getServerSideProps/all/*  /.netlify/functions/next_getServerSideProps_all_slug  200
/es/getServerSideProps/:id  /.netlify/functions/next_getServerSideProps_id  200
/es/getStaticProps/withFallback/:id  /.netlify/functions/next_getStaticProps_withFallback_id  200
/es/getStaticProps/withFallback/:slug/*  /.netlify/functions/next_getStaticProps_withFallback_slug  200
/es/getStaticProps/withFallbackBlocking/:id  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200
/es/getStaticProps/withRevalidate/withFallback/:id  /.netlify/functions/next_getStaticProps_withRevalidate_withFallback_id  200
/es/shows/:id  /.netlify/functions/next_shows_id  200
/es/shows/:params/*  /.netlify/functions/next_shows_params  200
/es/static/:id  /es/static/[id].html  200
/getServerSideProps/all  /.netlify/functions/next_getServerSideProps_all_slug  200
/getServerSideProps/all/*  /.netlify/functions/next_getServerSideProps_all_slug  200
/getServerSideProps/:id  /.netlify/functions/next_getServerSideProps_id  200
/getStaticProps/withFallback/:id  /.netlify/functions/next_getStaticProps_withFallback_id  200
/getStaticProps/withFallback/:slug/*  /.netlify/functions/next_getStaticProps_withFallback_slug  200
/getStaticProps/withFallbackBlocking/:id  /.netlify/functions/next_getStaticProps_withFallbackBlocking_id  200
/getStaticProps/withRevalidate/withFallback/:id  /.netlify/functions/next_getStaticProps_withRevalidate_withFallback_id  200
/shows/:id  /.netlify/functions/next_shows_id  200
/shows/:params/*  /.netlify/functions/next_shows_params  200
/static/:id  /en/static/[id].html  200"
`;
